File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofModels/CFBTeamAssets.swift
import Foundation

/// Process-wide cache of the `cfb_teams` reference table. CFB has 137 teams,
/// so cards resolve logos/colors from this cache instead of hardcoding a table.
@MainActor
public enum CFBTeamAssets {
    public private(set) static var byName: [String: CFBTeamReference] = [:]
    private static var nameByAlias: [String: String] = [:]

    public static var isLoaded: Bool { !byName.isEmpty }

    public static func install(_ teams: [CFBTeamReference]) {
        byName = Dictionary(uniqueKeysWithValues: teams.map { (normalize($0.teamName), $0) })
        var aliases: [String: String] = [:]
        for team in teams {
            let key = normalize(team.teamName)
            aliases[key] = key
            if let abbr = team.abbr, !abbr.isEmpty {
                aliases[normalize(abbr)] = key
            }
        }
        nameByAlias = aliases
    }

    public static func team(for name: String) -> CFBTeamReference? {
        let key = normalize(name)
        if let canonical = nameByAlias[key] {
            return byName[canonical]
        }
        return byName[key]
    }

    public static func abbr(for name: String) -> String {
        team(for: name)?.abbr?.uppercased() ?? fallbackAbbr(from: name)
    }

    public static func logo(for name: String, dark: Bool = false) -> String? {
        guard let team = team(for: name) else { return nil }
        return dark ? (team.logoDark ?? team.logo) : (team.logo ?? team.logoDark)
    }

    public static func colorHex(for name: String) -> (primary: String?, secondary: String?) {
        guard let team = team(for: name) else { return (nil, nil) }
        return (team.color, team.altColor)
    }

    public static func normalize(_ value: String) -> String {
        value
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: ".", with: "")
            .replacingOccurrences(of: "'", with: "")
            .replacingOccurrences(of: "  ", with: " ")
    }

    private static func fallbackAbbr(from name: String) -> String {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return "CFB" }
        return trimmed
            .split(separator: " ")
            .compactMap(\.first)
            .prefix(4)
            .map { String($0).uppercased() }
            .joined()
    }
}

File: wagerproof-android-native/core/models/src/main/kotlin/com/wagerproof/core/models/CFBTeamAssets.kt
package com.wagerproof.core.models

import java.util.Locale

/** Primary/secondary brand color hexes (Swift tuple `(primary, secondary)`). */
data class CFBTeamColorHex(val primary: String?, val secondary: String?)

/**
 * Process-wide cache of the `cfb_teams` reference table. CFB has 137 teams,
 * so cards resolve logos/colors from this cache instead of hardcoding a table.
 * iOS confines this to @MainActor; here an immutable snapshot swapped through
 * a @Volatile field gives the same install-once/read-anywhere semantics.
 */
object CFBTeamAssets {

    private data class Snapshot(
        val byName: Map<String, CFBTeamReference>,
        val nameByAlias: Map<String, String>,
    )

    @Volatile
    private var snapshot = Snapshot(emptyMap(), emptyMap())

    /** Normalized teamName → team reference. */
    val byName: Map<String, CFBTeamReference> get() = snapshot.byName

    val isLoaded: Boolean get() = snapshot.byName.isNotEmpty()

    fun install(teams: List<CFBTeamReference>) {
        val byName = teams.associateBy { normalize(it.teamName) }
        val aliases = mutableMapOf<String, String>()
        for (team in teams) {
            val key = normalize(team.teamName)
            aliases[key] = key
            val abbr = team.abbr
            if (!abbr.isNullOrEmpty()) {
                aliases[normalize(abbr)] = key
            }
        }
        snapshot = Snapshot(byName, aliases)
    }

    fun team(name: String): CFBTeamReference? {
        val snap = snapshot
        val key = normalize(name)
        snap.nameByAlias[key]?.let { canonical -> return snap.byName[canonical] }
        return snap.byName[key]
    }

    fun abbr(name: String): String =
        team(name)?.abbr?.uppercase(Locale.US) ?: fallbackAbbr(name)

    /** School name for display (e.g. "Kansas", "Texas Tech") — not the abbreviation. */
    fun displayName(name: String): String = team(name)?.teamName ?: name

    fun logo(name: String, dark: Boolean = false): String? {
        val team = team(name) ?: return null
        return if (dark) (team.logoDark ?: team.logo) else (team.logo ?: team.logoDark)
    }

    fun colorHex(name: String): CFBTeamColorHex {
        val team = team(name) ?: return CFBTeamColorHex(null, null)
        return CFBTeamColorHex(team.color, team.altColor)
    }

    fun normalize(value: String): String =
        value
            .trim()
            .lowercase(Locale.US)
            .replace(".", "")
            .replace("'", "")
            .replace("  ", " ")

    private fun fallbackAbbr(name: String): String {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return "CFB"
        return trimmed
            .split(" ")
            .mapNotNull { it.firstOrNull() }
            .take(4)
            .joinToString("") { it.toString().uppercase(Locale.US) }
    }
}
